plugins/fmhash/fmhash.c

```diff
@@ -0,0 +1,416 @@
+	hash_impl hashXX;
+	hash_wrapper_2 hash_wrapper_1_2;
+	hash_wrapper_3 hash_wrapper_2_3;
+	void* xhash;
+};
+
+/*
+ * Fowler–Noll–Vo hash 32 bit
+ * http://www.isthe.com/chongo/src/fnv/hash_32.c
+ */
+#if defined(__clang__)
+#pragma GCC diagnostic ignored "-Wunknown-attributes"
+#endif
+static uint32_t
+#if defined(__clang__)
+__attribute__((no_sanitize("unsigned-integer-overflow")))
+#endif
+fnv_32(const void* input, size_t len, uint32_t seed) {
+    unsigned char *bp = (unsigned char *)input;	/* start of buffer */
+
+    /*
+     * FNV-1 hash each octet in the buffer
+     */
+    size_t i;
+    for (i = 0; i < len; i++) {
+        /* multiply by the 32 bit FNV magic prime mod 2^32 */
+        seed += (seed<<1) + (seed<<4) + (seed<<7) + (seed<<8) + (seed<<24);
+
+        /* xor the bottom with the current octet */
+        seed ^= (uint32_t)*bp++;
+    }
+
+    /* return our new hash value */
+    return seed;
+}
+
+
+/*
+ * Modified Bernstein
+ * http://www.eternallyconfuzzled.com/tuts/algorithms/jsw_tut_hashing.aspx
+ */
+#if defined(__clang__)
+#pragma GCC diagnostic ignored "-Wunknown-attributes"
+#endif
+static uint64_t
+#if defined(__clang__)
+__attribute__((no_sanitize("unsigned-integer-overflow")))
+#endif
+djb_hash(const void* input, size_t len, uint32_t seed) {
+    const char *p = input;
+    uint64_t hash = 5381;
+    uint64_t i;
+    for (i = 0; i < len; i++) {
+        hash = 33 * hash ^ p[i];
+    }
+
+    return hash + seed;
+}
+
+/*Get 32 bit hash for input*/
+static rsRetVal
+hash32(const void* input, size_t len, uint32_t seed, hash_context_t* hcontext) {
+	DEFiRet;
+	uint64_t* xhash;
+	CHKmalloc(xhash = calloc(1, sizeof(uint64_t)));
+    #ifdef USE_HASH_XXHASH
+	    *xhash = XXH32(input, len, seed);
+    #else
+	    *xhash = fnv_32(input, len, seed);
+    #endif
+    hcontext->xhash = (void*)xhash;
+	iRet = RS_RET_OK;
+finalize_it:
+	RETiRet;
+}
+
+/*Get 64 bit hash for input*/
+static rsRetVal
+hash64(const void* input, size_t len, uint32_t seed, hash_context_t* hcontext) {
+	DEFiRet;
+	uint64_t* xhash;
+	CHKmalloc(xhash = calloc(1, sizeof(uint64_t)));
+	#ifdef USE_HASH_XXHASH
+        *xhash = XXH64(input, len, seed);
+    #else
+        *xhash = djb_hash(input, len, seed);
+    #endif
+    hcontext->xhash = (void*)xhash;
+    iRet = RS_RET_OK;
+finalize_it:
+	RETiRet;
+}
+
+#if !defined(__clang__)
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wmaybe-uninitialized"
+#endif
+static rsRetVal
+hash_wrapper2(struct svar *__restrict__ const sourceVal
+		, struct svar *__restrict__ const seedVal, hash_context_t* hcontext) {
+	DEFiRet;
+	int freeHashStr = 0, success = 0;
+    uint32_t seed = 0;
+    if(seedVal) {
+        seed = var2Number(seedVal, &success);
+        if (!success) {
+            DBGPRINTF("fmhash: hashXX(string, seed) didn't get a valid 'seed' limit, defaulting hash value to 0");
+            iRet = RS_RET_ERR;
+            FINALIZE;
+        }
+    }
+
+    char *hashStr = NULL;
```

Comment:
This is the source of the "maybe-unitialized". Please move it to after ling 164 and remove the disabling of the compiler diagnostic. We always try to compile clean with max warnings and disable them only if really, really there is no other way to solve the issue.

Reply:
My bad. Never realised this line was the issue. I have fixed it.

Reply:
I have become pretty used to this kind of issues - do them often enough myself ;-)

Reply:
Just for others coming across this thread: the root cause is that the compiler actually "allocates" the variable on the stack as soon as the function is entered. But is is uninitialized until that line is reached. In the concrete code here that was no issue due to the free-guard variable, but the compiler complains that this is dangerous practice as code may change in the future. It's often overlooked that C99 definition and assignment still means mem alloc on function entry.